sklearn/preprocessing/tests/test_polynomial.py

```diff
@@ -817,3 +1080,60 @@ def test_polynomial_features_behaviour_on_zero_degree():
         if sparse.issparse(output):
             output = output.toarray()
         assert_array_equal(output, np.ones((X.shape[0], 1)))
+
+
+def test_sizeof_LARGEST_INT_t():
+    # On Windows, scikit-learn is typically compiled with MSVC that
+    # does not support int128 arithmetic (at the time of writing):
+    # https://stackoverflow.com/a/6761962/163740
+    if sys.maxsize <= 2**32 or sys.platform == "win32":
```

Comment:
Actually maybe the problem above was caused by previously generated .c files. I need to try again from a clean build folder to make sure.

Reply:
I tried again on 32 bit linux on armv7 because I could find more recent python / clang versions for that platform and now the build works. But the resulting `_csr_polynomial_expansion` compiled extension is still broken:

```python-traceback
root@4af20c1dad16:/io# pytest -lxk test_sizeof_LARGEST_INT_t sklearn/preprocessing/tests/test_polynomial.py 
ImportError while loading conftest '/io/sklearn/conftest.py'.
sklearn/conftest.py:18: in <module>
    from sklearn.datasets import fetch_20newsgroups
sklearn/datasets/__init__.py:8: in <module>
    from ._base import load_breast_cancer
sklearn/datasets/_base.py:20: in <module>
    from ..preprocessing import scale
sklearn/preprocessing/__init__.py:38: in <module>
    from ._polynomial import PolynomialFeatures
sklearn/preprocessing/_polynomial.py:22: in <module>
    from ._csr_polynomial_expansion import (
E   ImportError: /io/sklearn/preprocessing/_csr_polynomial_expansion.cpython-310-arm-linux-gnueabihf.so: undefined symbol: __divti3
```

Reply:
Indeed I think we need to first check for `32bit` systems before eagerly trying `__int128` or `_BitInt(128)`. Probably the most portable way would be to check `sizeof(void*) == 8` right? I'll implement that soon.

Reply:
Indeed:

> If you compile for a 32-bit architecture like ARM, or x86 with -m32, no 128-bit integer type is supported with even the newest version of any of these compilers. So you need to detect support before using, if it's possible for your code to work at all without it.

from: https://stackoverflow.com/a/54815033/163740

I think we should also link to that SO question which is quite a good reference for this topic.

Reply:
Please also add a TODO note such that once all platform have C23 compilers, we should instead switch to the new way to do this:

> SO C23 will let you typedef unsigned _BitInt(128)  u128, modeled on [clang's feature](https://clang.llvm.org/docs/LanguageExtensions.html#extended-integer-types) originally called _ExtInt() which works even on 32-bit machines; see [a brief intro to it](https://stackoverflow.com/questions/61411865/how-do-you-use-clangs-new-custom-size-int-feature). Current GCC -std=gnu2x doesn't even support that syntax yet.

This will require a few years, but then it should work both for 32-bit and 64-bit target architectures.

Note that the clang + 32-bit target is no just a theoretical possibility: emscrypten / WASM / pyodide is a concrete platform that is built on that combo. I think @lesteve has the required tooling to test this PR on that target.

Reply:
Quick and dirty testing of this PR on Pyodide seems to indicate that the `test_sizeof_LARGEST_INT_t` fails but the other added tests pass:
```
>>> pytest.main(['--pyargs', 'sklearn.preprocessing', '-k', 'test_csr_polynomial_expansion_index_overflow_non_reg
ression or test_csr_polynomial_expansion_too_large_to_index or LARGEST_INT', '-v'])
============================= test session starts ==============================
platform emscripten -- Python 3.11.3, pytest-7.3.1, pluggy-1.0.0 -- /home/pyodide/this.program
cachedir: .pytest_cache
rootdir: /home/pyodide
collecting ... collected 1098 items / 1089 deselected / 9 selected                            
tests/test_polynomial.py::test_csr_polynomial_expansion_index_overflow_non_regression[True-True] PASSED [ 11%]
tests/test_polynomial.py::test_csr_polynomial_expansion_index_overflow_non_regression[True-False] PASSED [ 22%]
tests/test_polynomial.py::test_csr_polynomial_expansion_index_overflow_non_regression[False-True] PASSED [ 33%]
tests/test_polynomial.py::test_csr_polynomial_expansion_index_overflow_non_regression[False-False] PASSED [ 44%]
tests/test_polynomial.py::test_csr_polynomial_expansion_too_large_to_index[True-True] PASSED [ 55%]
tests/test_polynomial.py::test_csr_polynomial_expansion_too_large_to_index[True-False] PASSED [ 66%]
tests/test_polynomial.py::test_csr_polynomial_expansion_too_large_to_index[False-True] PASSED [ 77%]
tests/test_polynomial.py::test_csr_polynomial_expansion_too_large_to_index[False-False] PASSED [ 88%]
tests/test_polynomial.py::test_sizeof_LARGEST_INT_t FAILED               [100%]
=================================== FAILURES ===================================
__________________________ test_sizeof_LARGEST_INT_t ___________________________
    def test_sizeof_LARGEST_INT_t():
        # On Windows, scikit-learn is typically compiled with MSVC that
        # does not support int128 arithmetic (at the time of writing):
        # https://stackoverflow.com/a/6761962/163740
        if sys.maxsize <= 2**32 or sys.platform == "win32":
            expected_size = 8
        else:
            expected_size = 16
    
>       assert _get_sizeof_LARGEST_INT_t() == expected_size
E       assert 16 == 8
E        +  where 16 = _get_sizeof_LARGEST_INT_t()
/lib/python3.11/site-packages/sklearn/preprocessing/tests/test_polynomial.py:1094: AssertionError
=========================== short test summary info ============================
FAILED tests/test_polynomial.py::test_sizeof_LARGEST_INT_t - assert 16 == 8
================= 1 failed, 8 passed, 1089 deselected in 1.03s =================
```

> Note that the clang + 32-bit target is no just a theoretical possibility: emscripten / WASM / pyodide is a concrete platform that is built on that combo.

By the way I am not sure `__clang__` is defined in the Pyodide use case, I think `__EMSCRIPTEN__` is used instead.

Reply:
Okay so I have:
1. Added a check for `__EMSCRIPTEN__` as an alternative to `__clang__`
2. Updated the runtime test for largest integer type to account for the odd (but beneficial) exception that is emscripten

Reply:
Also when trying to build pydodide/emscripten for myself, I get the following error. I'm not sure what's going on here 🤷 :

<details>

```traceback

* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (Cython>=0.29.24, setuptools<=61.0, wheel)
* Getting dependencies for wheel...
Partial import of sklearn during the build process.
running egg_info
writing scikit_learn.egg-info/PKG-INFO
writing dependency_links to scikit_learn.egg-info/dependency_links.txt
writing requirements to scikit_learn.egg-info/requires.txt
writing top-level names to scikit_learn.egg-info/top_level.txt
reading manifest file 'scikit_learn.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
warning: no previously-included files matching '*' found under directory 'asv_benchmarks'
warning: no previously-included files matching '*' found under directory 'benchmarks'
warning: no previously-included files matching '*' found under directory 'build_tools'
warning: no previously-included files matching '*' found under directory 'maint_tools'
warning: no previously-included files matching '*' found under directory 'benchmarks'
warning: no previously-included files matching '*' found under directory '.binder'
warning: no previously-included files matching '*' found under directory '.circleci'
warning: no previously-included files found matching '.codecov.yml'
warning: no previously-included files found matching '.git-blame-ignore-revs'
warning: no previously-included files found matching '.mailmap'
warning: no previously-included files found matching '.pre-commit-config.yaml'
warning: no previously-included files found matching 'azure-pipelines.yml'
warning: no previously-included files found matching 'CODE_OF_CONDUCT.md'
warning: no previously-included files found matching 'CONTRIBUTING.md'
warning: no previously-included files found matching 'SECURITY.md'
warning: no previously-included files found matching 'PULL_REQUEST_TEMPLATE.md'
adding license file 'COPYING'
writing manifest file 'scikit_learn.egg-info/SOURCES.txt'
* Installing packages in isolated environment... (wheel)
* Building wheel...
Partial import of sklearn during the build process.
Traceback (most recent call last):
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/core/__init__.py", line 23, in <module>
    from . import multiarray
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/core/multiarray.py", line 10, in <module>
    from . import overrides
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/core/overrides.py", line 6, in <module>
    from numpy.core._multiarray_umath import (
ModuleNotFoundError: No module named 'numpy.core._multiarray_umath'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 221, in check_package_status
    module = importlib.import_module(package)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/__init__.py", line 141, in <module>
    from . import core
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/numpy/core/__init__.py", line 49, in <module>
    raise ImportError(msg)
ImportError:

IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!

Importing the numpy C-extensions failed. This error can happen for
many reasons, often due to issues with your setup or how NumPy was
installed.

We have compiled some common reasons and troubleshooting tips at:

    https://numpy.org/devdocs/user/troubleshooting-importerror.html

Please note and check the following:

  * The Python version is: Python3.11 from "/tmp/build-env-mnfj07nc/bin/python"
  * The NumPy version is: "1.24.2"

and make sure that they are the versions you expect.
Please carefully study the documentation linked above for further help.

Original error was: No module named 'numpy.core._multiarray_umath'

Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/pep517/in_process/_in_process.py", line 351, in <module>
    main()
  File "/usr/local/lib/python3.11/site-packages/pep517/in_process/_in_process.py", line 333, in main
    json_out['return_val'] = hook(**hook_input['kwargs'])
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/pep517/in_process/_in_process.py", line 249, in build_wheel
    return _build_backend().build_wheel(wheel_directory, config_settings,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/setuptools/build_meta.py", line 244, in build_wheel
    return self._build_with_temp_dir(['bdist_wheel'], '.whl',
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/setuptools/build_meta.py", line 229, in _build_with_temp_dir
    self.run_setup()
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/setuptools/build_meta.py", line 282, in run_setup
    self).run_setup(setup_script=setup_script)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/tmp/build-env-mnfj07nc/lib/python3.11/site-packages/setuptools/build_meta.py", line 174, in run_setup
    exec(compile(code, __file__, 'exec'), locals())
  File "setup.py", line 655, in <module>
    setup_package()
  File "setup.py", line 645, in setup_package
    check_package_status("numpy", min_deps.NUMPY_MIN_VERSION)
  File "setup.py", line 248, in check_package_status
    raise ImportError(
ImportError: numpy is not installed.
scikit-learn requires numpy >= 1.17.3.
Installation instructions are available on the scikit-learn website: http://scikit-learn.org/stable/install.html


ERROR Backend subproccess exited when trying to invoke build_wheel
```
</details>